src/main/java/com/salesforce/cdp/queryservice/util/QueryExecutor.java

```diff
@@ -128,9 +146,30 @@ protected Response getResponse(Request request) throws IOException {
         if (connection.getToken() != null && TokenHelper.isAlive(connection.getToken())) {
             return TokenHelper.getTokenWithUrl(connection.getToken());
         }
-        Token token = TokenHelper.getToken(connection.getClientInfo(), client);
-        connection.setToken(token);
-        Map<String, String> tokenMap = TokenHelper.getTokenWithUrl(token);
-        return tokenMap;
+        // todo: add a wrapper for retry mechanism
+        RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
+                .handle(TokenException.class)
+                .withMaxRetries(getMaxRetries(connection.getClientInfo()));
+        try {
+            return Failsafe.with(retryPolicy)
```

Comment:
Can you comment what failsafe will do here ?

Reply:
I have included the brief info about failsafe working